File: packages/typescript/service/src/classification.ts
/**
 * Event classification for v0.8 security hardening.
 */

const MALFORMED_FAILURE_TYPES = new Set([
  "unknown_capability",
  "streaming_not_supported",
  "internal_error",
]);

const HIGH_RISK_SIDE_EFFECTS = new Set([
  "write",
  "irreversible",
  "transactional",
]);

export function classifyEvent(
  sideEffectType: string | null,
  success: boolean,
  failureType: string | null,
): string {
  if (sideEffectType === null) {
    return "malformed_or_spam";
  }
  if (success) {
    if (HIGH_RISK_SIDE_EFFECTS.has(sideEffectType)) {
      return "high_risk_success";
    }
    return "low_risk_success";
  }
  if (failureType !== null && MALFORMED_FAILURE_TYPES.has(failureType)) {
    return "malformed_or_spam";
  }
  return "high_risk_denial";
}

File: packages/go/service/classification.go
package service

// ClassifyEvent assigns an event class to an audit entry based on
// side-effect type, success/failure, and failure type.
// Implements SPEC §6.8 event classification.
func ClassifyEvent(sideEffectType string, success bool, failureType string) string {
	if sideEffectType == "" {
		return "malformed_or_spam"
	}
	if success {
		if isHighRiskSideEffect(sideEffectType) {
			return "high_risk_success"
		}
		return "low_risk_success"
	}
	if isMalformedFailureType(failureType) {
		return "malformed_or_spam"
	}
	return "high_risk_denial"
}

func isHighRiskSideEffect(se string) bool {
	return se == "write" || se == "irreversible" || se == "transactional"
}

func isMalformedFailureType(ft string) bool {
	return ft == "unknown_capability" || ft == "streaming_not_supported" || ft == "internal_error"
}
